File: jput-core/src/main/java/cz/encircled/jput/io/file/FileSystemResultWriter.java
package cz.encircled.jput.io.file;

import cz.encircled.jput.io.TrendResultWriter;
import cz.encircled.jput.model.PerformanceTestRun;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static java.nio.file.StandardOpenOption.APPEND;
import static java.nio.file.StandardOpenOption.CREATE;

/**
 * @author Vlad on 21-May-17.
 */
public class FileSystemResultWriter implements TrendResultWriter {

    private final List<PerformanceTestRun> stack = new ArrayList<>();

    private final Object flushMutex = new Object();

    private Path target;

    public FileSystemResultWriter(String pathToFile) {
        this.target = Paths.get(pathToFile);
    }

    @Override
    public void appendTrendResult(PerformanceTestRun run) {
        synchronized (stack) {
            stack.add(run);
        }
    }

    @Override
    public void flush() {
        synchronized (flushMutex) {
            List<String> mapped;
            synchronized (stack) {
                mapped = stack.stream().map(this::toFileFormat).collect(Collectors.toList());
                stack.clear();
            }

            try {
                Files.write(target, mapped, StandardCharsets.UTF_8, APPEND, CREATE);
            } catch (IOException e) {
                throw new IllegalStateException("Failed to append to file", e);
            }
        }
    }

    private String toFileFormat(PerformanceTestRun run) {
        StringBuilder array = new StringBuilder();
        // TODO aggregate method runs?
        for (long l : run.runs) {
            if (array.length() > 0) {
                array.append(",");
            }
            array.append(l);
        }
        return String.join(";", Long.toString(run.executionId), array.toString(), run.testClass + "#" + run.testMethod);
    }

}

File: jput-core/src/main/java/cz/encircled/jput/io/file/FileSystemResultWriter.kt
package cz.encircled.jput.io.file

import cz.encircled.jput.io.TrendResultWriter
import cz.encircled.jput.model.PerformanceTestExecution
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardOpenOption.APPEND
import java.nio.file.StandardOpenOption.CREATE
import java.util.*

/**
 * @author Vlad on 21-May-17.
 */
class FileSystemResultWriter(pathToFile: String) : TrendResultWriter {

    private val stack = ArrayList<PerformanceTestExecution>()

    private val flushMutex = Any()

    private val target: Path = Paths.get(pathToFile)

    override fun appendTrendResult(execution: PerformanceTestExecution) {
        synchronized(stack) {
            stack.add(execution)
        }
    }

    override fun flush() {
        synchronized(flushMutex) {
            var mapped: List<String> = listOf()
            synchronized(stack) {
                mapped = stack.map(this::toFileFormat)
                stack.clear()
            }

            try {
                Files.write(target, mapped, StandardCharsets.UTF_8, APPEND, CREATE)
            } catch (e: IOException) {
                throw IllegalStateException("Failed to append to file", e)
            }

        }
    }

    private fun toFileFormat(execution: PerformanceTestExecution): String {
        // TODO aggregate method runs?
        val array = execution.runs!!.joinToString(",")
        return arrayOf(java.lang.Long.toString(execution.executionId), array, execution.testClass + "#" + execution.testMethod).joinToString(";")
    }

}
